### PR TITLE
[rpc] Add Rate Limiter for RPCs

### PR DESCRIPTION
**Pull request by AlexiaChen**
## Issue

https://github.com/harmony-one/harmony/issues/3696

- The current PR only supports BlockChain type RPCs, not other RPCs such as TxPool, Contract, etc. Because I don't know if restricting the flow of other RPCs will affect other products（ I'm not sure if I want to add rate limiter to all RPCs）
- I didn't add singleflight to `GetBlockByNumber` because I don't think it has much effect, because this needs to depend on the same key, but for explorer node, a large number of requests are not for the same key, because the block height is different. Unless a large number of requests are for blocks of the same height.

## Usage of rate limiter

- add two flags： `--rpc.ratelimiter` and `--rpc.ratelimit`. Or you can use it in config file, it all the same.
- the flag  `--rpc.ratelimiter` is default enabled, so you can ignore it
- the flag `-rpc.ratelimit` is used for control rate of requests, the value is the number of requests per sencond allowed with rate limiter.the default value is 100 (100/s).

so use rate limiter with shell, need to run as follows:

```bash
./harmony --rpc.ratelimiter --rpc.ratelimit=100
OR
./harmony
```

## Test

Use harmony explorer start a tons of requests, the real rate of requests is equal to config of `--rpc.ratelimit=<requests_per_second>`

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

  **NO**

### CODE REVIEW

**Review by rlan35 — APPROVED** [review]

**Review by LeoHChen — COMMENTED** [review]
LGTM in general, my question is about the default rate of 100/s. 

we need to spend some time tuning the rate.  How did you come up with this number?  Please work with Haodi to get some real-world value from the current mainnet.  We want to limit the request, but we also don't want to slow down the service.  Please check/analyze the existing explorer node, and get some data out of the load balancer and target group.

Thanks.

**Comment by AlexiaChen** [comment]
> 
> 
> LGTM in general, my question is about the default rate of 100/s.
> 
> we need to spend some time tuning the rate. How did you come up with this number? Please work with Haodi to get some real-world value from the current mainnet. We want to limit the request, but we also don't want to slow down the service. Please check/analyze the existing explorer node, and get some data out of the load balancer and target group.
> 
> Thanks.

the default rate is the result of my test on local, it's not real-world rate. If the speed reaches 1000/s, I find that the memory still keeps growing. Because the goroutine is always blocked by the rate limiter's wait method, the number of goroutines still keeps growing cumulatively.

**Comment by AlexiaChen** [comment]
> 
> 
> LGTM in general, my question is about the default rate of 100/s.
> 
> we need to spend some time tuning the rate. How did you come up with this number? Please work with Haodi to get some real-world value from the current mainnet. We want to limit the request, but we also don't want to slow down the service. Please check/analyze the existing explorer node, and get some data out of the load balancer and target group.
> 
> Thanks.

I ask Haodi the real data in the load balancer, It's 18000 requests / min in the real world, So, I set default value is 300/s(18000 / 60 = 300), That's alright ?

File: rpc/blockchain.go
Comment (JackyWYX) [inline rpc/blockchain.go:42]: What is the case for not limiting the rate?

File: rpc/blockchain.go
Comment (JackyWYX) [inline rpc/blockchain.go:229]: Can we refactor this into a new function?

**Review by JackyWYX — COMMENTED** [review]
Added some comments. BTW, have you considered migration of harmony config file? Also, it is recommended to have your config migration rebased on top of this PR after merge: https://github.com/harmony-one/harmony/pull/3681.

File: rpc/blockchain.go
Comment (AlexiaChen) [reply rpc/blockchain.go:229]: good point!! This urgent task about OOM makes me temporarily forget an important principle in software engineering - Do not repeat yourself. Thanks for your reminder!!

File: rpc/blockchain.go
Comment (AlexiaChen) [reply rpc/blockchain.go:42]: Now for the explorer node for the time being there is no unrestricted scenario, but I think this switch is still exposed as much as possible, this is a functional switch

**Pull request closed without merging**
